### DISCUSSION TITLE
Why is nixvim needed?

### DISCUSSION THREAD

**Discussion by refaelsh** [nix-community/nixvim#2017] [Q&A] [answered]
I think `nixvim` is a cool idea.
But, why is it needed? What advantage does it have over simply using `~/.config/nvim/init.lua`?
Or maybe even creating `~/.config/nvim/init.lua` via home-manager.

**Comment by traxys** [comment]
It allow to bundle with your neovim executable the plugins (pinned to the nixpkgs version you use), and the executables like LSP that are required.
Using nix modules also allows for a thighter intergration with you other programs, and allows to modify the config for each project easily.

**Reply by refaelsh** [reply]
> Using nix modules also allows for a thighter intergration with you other programs

Can you please provide an example?

As for the other points you raised, every one of them can be achieved with init.lua.

**Reply by MattSturgeon** [reply]
Well, fundamentally all nixvim does is produce an `init.lua`, packaged using [`pkgs.wrapNeovimUnstable`](https://github.com/NixOS/nixpkgs/blob/31cbc1e205512a73086cc631ecbd70aa9e7d33ef/pkgs/applications/editors/neovim/wrapper.nix). So from that perspective, of course you can achieve similar results manually or with some other tool.

At the end of the day, nixvim is not _needed_ and may users configure neovim without it.

Nixvim is geared toward people who prefer configuring options using the [nixos module system](https://nix.dev/tutorials/module-system/index.html). If you prefer configuring directly in lua, then nixvim probably isn't for you.

As a (somewhat contrived) example, you could have some config be conditional on some other config:

```nix
{ config, lib, ... }: {
  opts.relativenumber = config.plugins.treesitter.enable;
}
```

Nixvim also provides a nix-module representation of many plugin's `setup()` function options, which many users who prefer using nix to lua appreciate. This is our biggest time-sink while maintaining nixvim though :sweat_smile:

**Comment by khaneliman** [comment, marked as answer]
- It  includes tests that help identify missing dependencies for plugins before you use your config
- Sanity check warnings/assertions for your config if you try and do something incorrectly or that can lead to unexpected behavior. 
- People who are comfortable with nix but dont like writing lua
- Allows reducing amount of code needed to be written by focusing on configuration instead of how to setup/call everything.

**Comment by refaelsh** [comment]
Thank you all who answered. That helped me get my head together.

P.S.
I will give nixvim a try. Definitly.
